Sources/SwiftFusion/Core/Dictionary+Differentiable.swift

```diff
@@ -0,0 +1,52 @@
+/// This file makes `Dictionary` differentiable.
+///
+/// Note: This will eventually be moved into the Swift standard library. Once it is in the
+/// standard library, we can delete it from this repository.
+
+/// Implements the `Differentiable` requirements.
+extension Dictionary: Differentiable where Value: Differentiable {
+  public typealias TangentVector = Dictionary<Key, Value.TangentVector>
+  public mutating func move(along direction: TangentVector) {
+    for (componentKey, componentDirection) in direction {
+      func fatalMissingComponent() -> Value {
+        fatalError("missing component \(componentKey) in moved Dictionary")
+      }
+      self[componentKey, default: fatalMissingComponent()].move(along: componentDirection)
+    }
+  }
+  public var zeroTangentVector: TangentVector { mapValues { _ in .zero } }
+}
+
+/// Implements the `AdditiveArithmetic` requirements.
+extension Dictionary: AdditiveArithmetic where Value: AdditiveArithmetic {
+  public static func + (_ lhs: Self, _ rhs: Self) -> Self {
+    lhs.merging(rhs, uniquingKeysWith: +)
+  }
+  public static func - (_ lhs: Self, _ rhs: Self) -> Self {
+    lhs.merging(rhs.mapValues { .zero - $0 }, uniquingKeysWith: +)
+  }
+  public static var zero: Self { [:] }
+}
+
+/// Provides some differentiable methods for manipulating `Dictionary`.
+///
+/// Note: Once the differentiable `Dictionary` is moved into the standard library, the standard
+/// `Dictionary` methods will be differentiable and you won't have to use special differentiable
+/// methods to manipulate `Dictionary`.
+extension Dictionary where Value: Differentiable {
+  /// Returns the value with `key`.
+  ///
+  /// Precondition: `self` contains an entry with key `key`.
+  @differentiable
+  public func differentiableSubscript(_ key: Key) -> Value {
```

Comment:
So the user need to call this instead of `dict[ key ]` (the normal subscript), right?

Reply:
Yes. The compiler should give nice errors pointing at any uses of `dict[ key ]` in differentiable functions, so it should be pretty easy to fix whenever the user forgets.

Eventually we'll be able to support derivatives of `dict[ key ]`, but there are a few things blocking them from working right now.